File: src/server/metadata.ts
/**
 * SOP metadata extraction (Sprint 6) — ported from `services/metadata.py`.
 *
 * Pulls document-level metadata (title, department, version, effective date,
 * owner) from the header region using labelled-field heuristics, plus topical
 * tags derived from keyword presence.
 */

import type { SopMetadata } from "@/lib/types";

type MetaField = "title" | "department" | "version" | "effective_date" | "owner";

const FIELD_PATTERNS: [MetaField, RegExp][] = [
  ["department", /^\s*department\s*[:\-]\s*(.+)$/i],
  ["version", /^\s*(?:version|revision|rev)\s*[:\-]?\s*v?\s*([\w.]+)$/i],
  ["effective_date", /^\s*(?:effective(?:\s+date)?|date)\s*[:\-]\s*(.+)$/i],
  ["owner", /^\s*(?:owner|author|approved by|prepared by)\s*[:\-]\s*(.+)$/i],
  ["title", /^\s*(?:title|sop title|document)\s*[:\-]\s*(.+)$/i],
];

const TAG_KEYWORDS: [string, string][] = [
  ["refund", "refunds"],
  ["onboard", "onboarding"],
  ["complian", "compliance"],
  ["security", "security"],
  ["approval", "approvals"],
  ["escalat", "escalation"],
  ["privacy", "privacy"],
  ["payment", "payments"],
  ["hr ", "hr"],
  ["audit", "audit"],
];

export function extractMetadata(text: string): SopMetadata {
  const meta: SopMetadata = { tags: [] };
  const lines = text.split(/\r?\n/);
  const header = lines.slice(0, 25);

  for (const line of header) {
    for (const [field, pattern] of FIELD_PATTERNS) {
      const match = pattern.exec(line);
      if (match && !meta[field]) {
        meta[field] = match[1].trim();
      }
    }
  }

  if (!meta.title) {
    const firstNonEmpty = lines.find((l) => l.trim());
    if (firstNonEmpty) meta.title = firstNonEmpty.trim().slice(0, 120);
  }

  const lowered = text.toLowerCase();
  meta.tags = [
    ...new Set(
      TAG_KEYWORDS.filter(([needle]) => lowered.includes(needle)).map(
        ([, tag]) => tag
      )
    ),
  ].sort();

  return meta;
}

File: backend/app/services/metadata.py
"""SOP metadata extraction (Sprint 6).

Pulls document-level metadata (title, department, version, effective date,
owner) from the header region of an SOP using labelled-field heuristics, plus
a set of topical tags derived from keyword frequency.
"""

from __future__ import annotations

import re

from app.models.schemas import SopMetadata

_FIELD_PATTERNS = {
    "department": re.compile(r"^\s*department\s*[:\-]\s*(.+)$", re.IGNORECASE),
    "version": re.compile(
        r"^\s*(?:version|revision|rev)\s*[:\-]?\s*v?\s*([\w.]+)$", re.IGNORECASE
    ),
    "effective_date": re.compile(
        r"^\s*(?:effective(?:\s+date)?|date)\s*[:\-]\s*(.+)$", re.IGNORECASE
    ),
    "owner": re.compile(
        r"^\s*(?:owner|author|approved by|prepared by)\s*[:\-]\s*(.+)$",
        re.IGNORECASE,
    ),
    "title": re.compile(
        r"^\s*(?:title|sop title|document)\s*[:\-]\s*(.+)$", re.IGNORECASE
    ),
}

_TAG_KEYWORDS = {
    "refund": "refunds",
    "onboard": "onboarding",
    "complian": "compliance",
    "security": "security",
    "approval": "approvals",
    "escalat": "escalation",
    "privacy": "privacy",
    "payment": "payments",
    "hr ": "hr",
    "audit": "audit",
}


def extract_metadata(text: str) -> SopMetadata:
    meta = SopMetadata()
    header = text.splitlines()[:25]

    for line in header:
        for field, pattern in _FIELD_PATTERNS.items():
            match = pattern.match(line)
            if match and not getattr(meta, field):
                setattr(meta, field, match.group(1).strip())

    # Fall back to the first non-empty line as the title.
    if not meta.title:
        for line in text.splitlines():
            if line.strip():
                meta.title = line.strip()[:120]
                break

    lowered = text.lower()
    tags = sorted(
        {tag for needle, tag in _TAG_KEYWORDS.items() if needle in lowered}
    )
    meta.tags = tags
    return meta
